
<html>
<head>
<title>DRIP tandem mass spectrometry analysis software</title>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-72240246-2', 'auto');
  ga('send', 'pageview');

</script>
<link href="./drip.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<center>
<img  src="dripSearchFlowchart.png" usemap="#commands-files" width="60%">
<!-- <map name="commands-files"> -->
<!--   <area shape="rect" coords="253,135,349,169" alt="dripDigest" href="dripDigest.html"/> -->
<!--   <area shape="rect" coords="260,247,342,278" alt="dripTrain" -->
<!--   href="dripTrain.html"/> -->
<!--   <area shape="rect" coords="260,247,342,278" alt="dripSearch" href="dripSearch.html"/> -->
<!-- </map> -->
</center>

<p>
  The DRIP Toolkit utilizes a dynamic Bayesian network (DBN) for
  Rapid Identification of Peptides (DRIP) in tandem
  mass spectra.  Given an observed spectrum, DRIP scores a peptide by aligning the peptide's theoretical spectrum and the observed spectrum, i.e., computing the most probable sequence of insertions (spurious 
observed peaks) and deletions (missing theoretical peaks).  DBN inference is efficiently performed utilizing the
  <a href="https://melodi.ee.washington.edu/gmtk/">Graphical Models
  Toolkit (GMTK)</a>, which allows easy alteration of the model.  If
  you use the DRIP toolkit in your research, please
  cite:</p><blockquote>John T. Halloran, Jeff
  A. Bilmes, and William S.
  Noble. <a href="http://auai.org/uai2014/proceedings/individuals/260.pdf">"Learning
  Peptide-Spectrum Alignment Models for Tandem Mass
  Spectrometry"</a>. <em>Thirtieth Conference on Uncertainty in
  Artificial Intelligence (UAI 2014)</em>. AUAI, Quebic City, Quebec
  Canada, July 2014.</blockquote>

<p />
<a href="https://github.com/melodi-lab/dripToolkit" title="Download the DRIP Toolkit" class="sfdl">
    <span>
        <b>Download the DRIP Toolkit</b>
    </span>
</a>
<div style="clear:both;"></div>
<p/>
The toolkit is provided as a set of python modules modules:</p>

<table>

<tr>
<td colspan="2"><b>Primary modules</b></td>
</tr>
<tr>
<td>
<a href="dripDigest.html">dripDigest.py</a></td>
<td>
Create an index of all peptides in a fasta file, for use in subsequent
calls to <code>dripSearch.py</code>.</td>
</tr>

<tr>
<td>
<a href="dripTrain.html">dripTrain.py</a></td>
<td>
Train DRIP model parameters given a set of high-confidence PSMs.
Only applicable for low-resolution MS2 spectra.</td>
</tr>

<tr>
<td>
<a href="dripSearch.html">dripSearch.py</a></td>
<td>
Search a collection of spectra against a sequence database, provided
as the output of <code>dripDigest.py</code>, returning a collection of
peptide-spectrum matches (PSMs).</td>
</tr>

<!-- <tr> -->
<!-- <td colspan="2"><b>Utilities</b></td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="assign-confidence.html">assign-confidence</a></td> -->
<!-- <td> -->
<!-- Assign statistical confidence measures to each PSM in a given set.</td></tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="generate-peptides.html">generate-peptides</a></td> -->
<!-- <td> -->
<!-- Extract from a given set of protein sequences a list of target and -->
<!-- decoy peptides fitting the specified criteria.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="get-ms2-spectrum.html">get-ms2-spectrum</a></td> -->
<!-- <td> -->
<!-- Extract one or more fragmentation spectra, specified by scan number, from -->
<!-- an <code>MS2</code> file.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td><a href="hardklor.html">hardkl&ouml;r</a></td> -->
<!-- <td>Identify isotopic distributions from high-resolution mass spectra.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td><a href="make-pin.html">make-pin</a></td> -->
<!-- <td>Given a set of search results files, generate a pin file for input to -->
<!-- <code>crux percolator</code></td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="predict-peptide-ions.html">predict-peptide-ions</a></td> -->
<!-- <td> -->
<!-- Given a peptide and a charge state, predict the m/z values of the -->
<!-- resulting fragment ions.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="print-processed-spectra.html">print-processed-spectra</a></td> -->
<!-- <td> -->
<!-- Process spectra as for scoring xcorr and print the results to a file.</td></tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="subtract-index.html">subtract-index</a></td> -->
<!-- <td>Subtract an index file from another both generated by tide-index.  -->
<!-- It is used to create disjoint pepitde indexes.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="xlink-assign-ions">xlink-assign-ions</a></td> -->
<!-- <td>Given a spectrum and a pair of cross-linked peptides, assign theoretical -->
<!-- ion type labels to peaks in the observed spectrum.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="xlink-score-spectrum">xlink-score-spectrum</a></td> -->
<!-- <td>Takes a defined cross-linked peptide, a spectra file, and a scan number, -->
<!-- and will calculate the XCorr score a number of different ways.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- version</td> -->
<!-- <td> -->
<!-- Print the Crux version number to standard output, then exit.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td colspan="2"><b>Utilities for processing tab-delimited text files</b> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="extract-columns.html">extract-columns</a></td> -->
<!-- <td>Print specified columns from a tab-delimited file.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="extract-rows.html">extract-rows</a></td> -->
<!-- <td>Print specified rows from a tab-delimited file.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="stat-column.html">stat-column</a></td> -->
<!-- <td>Collect summary statistics from a column in a tab-delimited file.</td> -->
<!-- </tr> -->

<!-- <tr> -->
<!-- <td> -->
<!-- <a href="sort-by-column.html">sort-by-column</a></td> -->
<!-- <td>Sort a tab-delimited file by a column.</td> -->
<!-- </tr> -->

</table>

<h2>File formats</h2>

<table id="fileFormatsTable" border="1">

<tr>
<td>&nbsp;</td>
<td>dripDigest</td>
<td>dripTrain</td>
<td>dripSearch</td>
</tr>

<tr>
<td>
<a href="ms2-format.html">MS2</a>
</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
</tr>

<tr>
<td>
Tab-delimited training PSM file
(<a href="sample.psm">sample</a>)
</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>
GMTK learned parameters
</td>
<td>&nbsp;</td>
<td>out</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="fasta-format.html">FASTA</a>
(<a href="sample.fasta">sample</a>)
</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>


<tr>
<td>
database index
</td>
<td>out</td>
<td>&nbsp;</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="txt-format.html">Tab-delimited text</a>
(<a href="dripSearch-sampleOutput.txt">sample</a>)
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
</tr>

</table>

<h2><a href="documentation.html">DRIP Toolkit documentation</a></h2>
<!-- <li> -->
<!-- <a href="install.html">Installation</a></li> -->

<!-- <li> -->
<!-- <a href="start-tutorial.html">DRIP toolkit tutorial</a></li> -->
<!-- </ul> -->

<hr></hr>

<p>
Please send comments and questions to halloj3@uw.edu.</p>

</blockquote>
</body>
</html>
